Complete the support for try_table.

sjrd opened the pull request:
(Motivation: the Scala.js Wasm backend emits `try_table`s, and we would like to pipe our output to `wasm-opt`.)

This includes adding interpreter support for `exnref` values. Since they must carry a `Name tag` in addition to `Literals payload`, they did not fit into existing `GCData`. Therefore, we add another possible type in the big `union` for `ExnData`.

Alternatively, we could make a non-`funcref` `Name` value a valid choice for a single `Literal`, but that would require inventing a fake `Type` for such literals, such as `Type::tag`, which would have no spec equivalent.

Other than the above, the changes are straightforward. They are inspired from existing handling of `Try` and `Break`.

Currently WiP:

* ~~The new tests in `local-subtyping.wast` fail. Some pass somewhere is refining the types of blocks that are the target of catch clauses in an incorrect way. The obvious candidates--`finalize`, `operateOnScopeNameUsesAndSentTypes`--seem to be doing the right thing, so I am missing something somewhere.~~
* More existing tests for legacy eh should be translated to the new eh. Is the approach taken for the tests translated so far appropriate?

I would appreciate some guidance about the above questions. ~~Can someone point me in the right direction for the first one, in particular?~~

Also, is the approach taken for `ExnData` acceptable? Should I do it differently?

Thanks in advance

kripken commented on src/literal.h:
```diff
@@ -732,6 +738,17 @@ struct GCData {
+  Literals payload;
+
+  ExnData(Name tag, Literals payload) : tag(tag), payload(payload) {}
+};
```
If we identify exception tags by their name, how does that handle cross-module catches? I mean
```wat
(module $A
  (tag $foo (import "a" "b"))
  (func $throw (export "A.throw")
    (throw a $foo)
  )
)
(module $B
  (tag $bar (import "a" "b"))
  (func $catch
    (block $out
      (try_table (catch $bar $out)
        (call $A.throw)
      )
    )
  )
)
```
The same tag is imported into two different modules, but with different internal names. I think the catch here would not properly catch?

sjrd replied:
Huh, that's a good point. It hadn't crossed my mind. I'll have to think about it tomorrow. I'll probably need some actual run-time representation of a tag, after all.

sjrd replied:
I've investigated a bit, and I think this is a larger problem than what this PR tackles. It seems the existing support for `try` has the same issue: the thrown `WasmException::tag` is also the `Name` of the tag in one module, and if another module tries to catch the exception, the `Name`s might not match.

This seems to be an issue only for the `shell-interface.h`, which is the only one that actually implements `ExternalInterface::throwException` as throwing the `WasmException`.

AFAICT, the same issue also exists for `ref.func`s: a `Literal` of a function type holds a simple `Name`, and it does not seem like any effort is made to translate those `Name`s when they cross module boundaries. Later, a `call_ref` looks up that `Name` in the current module's function map:
https://github.com/WebAssembly/binaryen/blob/23a1a1aacc0dd14767ca8c53a034b7c6bb4acf52/src/wasm-interpreter.h#L4139

kripken replied:
Good points. RefFunc should have a reference to the actual function, not the name, and likewise for these others. Yeah, I agree this is a larger issue than this PR, and I guess we missed it because we don't currently test cross-module calls. Sounds good to ignore it here, but maybe add something like `TODO: handle cross-module calls using something other than a Name here`.

kripken commented on src/wasm/literal.cpp:
```diff
@@ -147,7 +156,6 @@ Literal::Literal(const Literal& other) : type(other.type) {
-        case HeapType::exn:
           WASM_UNREACHABLE("invalid type");
         case HeapType::string:
           WASM_UNREACHABLE("TODO: string literals");
```
About 15 lines below here, there is the destructor `~Literal`. It needs code to destroy `exnData`, parallel to `gcData`.

sjrd replied:
Oops, thanks. I've lived to long among languages with a GC; I lost some reflexes. :laughing: 

Fixed.

kripken reviewed (commented):
Thanks for the PR!

Some initial comments on the code. I haven't read the tests yet.

kripken commented:
Aside from those comments the code looks right to me - nice work!

sjrd commented:
Ah ah! I found the issue with the refined block types. There was some code missing from `ReFinalize::visitTryTable`. All the existing tests now pass locally.

kripken commented on test/lit/passes/dce-eh.wast:
```diff
@@ -0,0 +1,145 @@
+;; RUN: wasm-opt %s --dce -all -S -o - | filecheck %s
+
+;; If either try body or catch body is reachable, the whole try construct is
+;; reachable
```
```suggestion
;; reachable.
```

kripken commented on test/lit/passes/dce-eh.wast:
```diff
@@ -0,0 +1,145 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $throw
+    ;; All these wrapping expressions before 'throw' will be dce'd
```
```suggestion
    ;; All these wrapping expressions before 'throw' will be dce'd.
```

kripken commented on test/lit/passes/local-subtyping.wast:
```diff
@@ -568,4 +571,48 @@
+  ;; CHECK-NEXT: )
+  (func $try-table-catch-result
+    (drop
+      ;; Must not be refined to nullref
```
```suggestion
      ;; Must not be refined to nullref.
```

kripken commented on test/lit/passes/local-subtyping.wast:
```diff
@@ -568,4 +571,48 @@
+  ;; CHECK-NEXT: )
+  (func $try-table-ref
+    (drop
+      ;; Must not be refined to nullexnref
```
```suggestion
      ;; Must not be refined to nullexnref.
```

kripken commented on test/lit/passes/local-subtyping.wast:
```diff
@@ -568,4 +571,48 @@
+        (ref.null exn)
+      )
+    )
+  )
```
Is this test needed? It seems very similar to the one above it.

sjrd replied:
One tests the case where the value is sent because it's part of the exception's payload. The other is when it's the `_ref` that generates an `exnref`. It's actually the latter I ran into in a real context.

To me those are distinct causes, which deserve distinct tests. But if you disagree, I'm happy to remove one of them.

kripken replied:
Ah, good point. Yes, those are distinct causes, let's keep the new test.

kripken commented on test/spec/return_call_eh-legacy.wast:
```diff
@@ -0,0 +1,35 @@
+;; Test the combination of 'return_call' with legacy exception handling
```
```suggestion
;; Test the combination of 'return_call' with legacy exception handling.
```

kripken reviewed (approved):
Nice work, looks good!

Let's also get feedback from @aheejin before landing.

@tlively may also want to look at the spec test here.

kripken commented:
Oh, one thing to consider here is that any new tests that are added get quickly looked at by our fuzzer. Until we are certain the fuzzer is ready for try_table, please add the new tests here:

https://github.com/WebAssembly/binaryen/blob/fb6ead80296471276f4cee05f920e6fe8aba67c5/scripts/fuzz_opt.py#L355-L358

(You can also run that script locally (instructions at the top) to see if it finds anything, to check if that is needed.)

tlively reviewed (commented):
Spec tests LGTM, thanks!

sjrd commented:
Thank you for the reviews. I believe I have addressed all the points that were raised.

I also completed the translation of `-eh-legacy` lit tests to the new EH under `-eh` tests. That uncovered one incorrect optimization in `CodeFolding`, and two missed opportunities in `CodePushing` and `Vacuum`. If fixed those in the process.

I don't consider this WiP anymore. If you're happy with it, I think it's complete.

---

To be fully transparent, there are two things that were not done here:

There is a TODO in `Vacuum.cpp` for an optimization that we do for `try` but not yet for `try_catch`. I left the translation of the tests for it in as "progression" tests, so it will be easy to adapt if/when we add that optimization.

Possibly more important: I discovered `try_table` is still not supported in `Flatten.cpp`. I did not address this. I don't think it's *possible* to address it and keep the restrictions on the target IR as described in `ir/flat.h`. The problem is that the target blocks for the `catch`es *must* necessarily have result types to capture the exceptions' payloads and/or the `exnref`s. There is no way to "configure" the `try_table` to store these in locals instead. I'm not sure how important the Flat IR is? This missing piece was not triggered when I applied `wasm-opt` to the Scala.js test suite, so I am not directly affected, but perhaps others will be?

sjrd commented:
> (You can also run that script locally (instructions at the top) to see if it finds anything, to check if that is needed.)

Looks like I found an issue but with `f32.store_f16` instead :smile: . I'll see how I'm supposed to report it tomorrow or later this weekend.

kripken commented:
@sjrd 

> I discovered try_table is still not supported in Flatten.cpp

Yeah, that is a larger issue that we need to think about. I agree with your analysis. We can land this without that.

> Looks like I found an issue but with f32.store_f16 instead 😄 . I'll see how I'm supposed to report it tomorrow or later this weekend.

Ah, no need: that was a recent landing that has since been fixed by ignoring f16 in the fuzzer in #6822. Merging latest main should avoid it for you.

kripken commented on scripts/fuzz_opt.py:
```diff
@@ -350,10 +350,20 @@ def is_git_repo():
     'exception-handling.wast',
+    'global-effects.wast',
+    'local-subtyping.wast',
+    'renamings.wat',
```
These three tests are not strictly wasm-EH tests, though perhaps it is fine to disable them for now as they are not terribly important. Maybe put them at the end with a separate comment like `These contain parts with new EH. TODO: split those parts out if we see that fuzzing new EH is far out`. (but my hope is that it is not far out).

kripken commented on src/ir/effects.h:
```diff
@@ -472,6 +480,24 @@ class EffectAnalyzer {
     }
 
+    static void doStartTryTable(InternalAnalyzer* self, Expression** currp) {
+      TryTable* curr = (*currp)->cast<TryTable>();
```
```suggestion
      auto* curr = (*currp)->cast<TryTable>();
```

kripken replied:
(Looks like the code above that this is based on has the same issue, but let's add the new code in the proper form. We can separately improve the other code - if you feel like it, that could be done in this PR too.)

kripken commented on src/ir/effects.h:
```diff
@@ -472,6 +480,24 @@ class EffectAnalyzer {
+    }
+
+    static void doEndTryTable(InternalAnalyzer* self, Expression** currp) {
+      TryTable* curr = (*currp)->cast<TryTable>();
```
```suggestion
      auto* curr = (*currp)->cast<TryTable>();
```

kripken commented on src/passes/CodeFolding.cpp:
```diff
@@ -347,11 +347,12 @@ struct CodeFolding : public WalkerPass<ControlFlowWalker<CodeFolding>> {
+        // the move if 'outOf' contains a 'try' or 'try_table' anywhere in it.
+        // This is a conservative approximation because there can be cases that
+        // 'try'/'try_table' is within the expression that may throw so it is
+        // safe to take the expression out.
```
```suggestion
        // safe to take the expression out.
        // TODO: optimize this check to avoid two FindAlls.
```

kripken commented on src/passes/Vacuum.cpp:
```diff
@@ -435,6 +435,22 @@ struct Vacuum : public WalkerPass<ExpressionStackWalker<Vacuum>> {
+    // though. Instead, we should turn it into an unconditional `br` to the
+    // target of the `catch_all`. Also, for a `catch_all_ref`, we cannot do that
+    // optimization at all since we would not know how to create the correct
+    // `exnref`.
```
```suggestion
    // `exnref`. Note: This optimization may make more sense in RemoveUnusedBrs.
```

kripken reviewed (commented):
Nice work! Here are a few minor comments on the code. I'll look at the tests next.

Btw, please don't squash and force-push, if possible: This is a large PR, and seeing incremental small commits makes it easier to do incremental reviews. We will squash all the commits when landing anyhow.

kripken commented on test/lit/passes/coalesce-locals-eh.wast:
```diff
@@ -0,0 +1,86 @@
+
+ (tag $e)
+
+ (tag $any (param (ref any)))
```
Indentation here looks different than the function before. Also, please put the tags first, before functions.

kripken commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,290 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-success (result i32)
+    (block $tryend
```
Please add a comment explaining what is tested here.

sjrd commented:
> Btw, please don't squash and force-push, if possible: This is a large PR, and seeing incremental small commits makes it easier to do incremental reviews. We will squash all the commits when landing anyhow.

Oops, sorry. Different habits in different repos. I should have asked first.

This time I added a commit addressing the latest batch of comments.

kripken commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $br-in-catch
+    ;; When the try_table removed, the removal of the implicit branch to the
```
```suggestion
    ;; When the try_table is removed, the removal of the implicit branch to the
```

kripken commented:
No worries, yeah, there isn't really a standard way of doing these things on github...

Just one last minor comment from me. Otherwise, looks good, except for @aheejin taking a look, and also @sjrd please see that at least a few thousand fuzzer iterations pass without issue here.

aheejin commented:
Thanks for the work! Sorry I'm late; I'll be posting comments shortly.

aheejin commented on src/wasm-interpreter.h:
```diff
@@ -47,6 +47,7 @@
 namespace wasm {
 
 struct WasmException {
+  // TODO: handle cross-module calls using something other than a Name here.
   Name tag;
   Literals values;
 };
```
Now that we have `exnref`, this can simply be
```cpp
  struct WasmException {
    WasmException(Literal exn) : exn(exn) {}
    Literal exn;
  };
```

sjrd replied:
I tried to something like that, but I'm running into issues that I have a hard time troubleshooting. I get the following errors when running the `spec` tests:
```
executing:  /localhome/doeraene/projects/binaryen/bin/wasm-shell /localhome/doeraene/projects/binaryen/test/spec/exception-handling-legacy.wast
('run_command failed (-11)', '0 BUILDING MODULE [line: 1]\n1 CHECKING [line: 288]\n----------------------------------------\nCAUGHT EXCEPTION\n2 CHECKING [line: 289]\n----------------------------------------\nCAUGHT EXCEPTION\n3 CHECKING [line: 290]\n4 CHECKING [line: 291]\n')
Traceback (most recent call last):
  File "/localhome/doeraene/projects/binaryen/./check.py", line 214, in run_spec_tests
    actual = run_spec_test(wast)
  File "/localhome/doeraene/projects/binaryen/./check.py", line 190, in run_spec_test
    output = support.run_command(cmd, stderr=subprocess.PIPE)
  File "/localhome/doeraene/projects/binaryen/scripts/test/support.py", line 193, in run_command
    raise Exception(('run_command failed (%s)' % code, out + str(err or '')))
Exception: ('run_command failed (-11)', '0 BUILDING MODULE [line: 1]\n1 CHECKING [line: 288]\n----------------------------------------\nCAUGHT EXCEPTION\n2 CHECKING [line: 289]\n----------------------------------------\nCAUGHT EXCEPTION\n3 CHECKING [line: 290]\n4 CHECKING [line: 291]\n')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/localhome/doeraene/projects/binaryen/./check.py", line 408, in <module>
    sys.exit(main())
  File "/localhome/doeraene/projects/binaryen/./check.py", line 391, in main
    TEST_SUITES[test]()
  File "/localhome/doeraene/projects/binaryen/./check.py", line 220, in run_spec_tests
    shared.fail_with_error(str(e))
  File "/localhome/doeraene/projects/binaryen/scripts/test/shared.py", line 334, in fail_with_error
    raise Exception(msg)
Exception: ('run_command failed (-11)', '0 BUILDING MODULE [line: 1]\n1 CHECKING [line: 288]\n----------------------------------------\nCAUGHT EXCEPTION\n2 CHECKING [line: 289]\n----------------------------------------\nCAUGHT EXCEPTION\n3 CHECKING [line: 290]\n4 CHECKING [line: 291]\n')
```
I pushed my attempt so far at https://github.com/sjrd/binaryen/compare/full-support-try-table...sjrd:binaryen:exnref-in-wasmexception

kripken replied:
@sjrd You can run individual spec tests using e.g.
```
bin/wasm-shell test/spec/exception-handling-legacy.wast
```

aheejin replied:
I'm attending some seminar Tue-Wed, so I can't take a look now, but will take a look soon, unless it's solved by then. But I think this can be done by a follow-up too.

sjrd replied:
> @sjrd You can run individual spec tests using e.g.
> 
> ```
> bin/wasm-shell test/spec/exception-handling-legacy.wast
> ```

Thanks, that helps a bit. At least now I know I'm running into a segfault during catch handling inside `visitTry`. It's a bit of a heisenbug: depending on how much printing I do, I get further or not. I think I'm out of my depth in terms of understanding how C++ handles exception structs that have a destructor. :confused:

kripken replied:
No worries @sjrd . Let's leave this part to a followup then, as @aheejin suggested. It does sound useful but it may not be trivial to do, and is not urgent IIUC.

kripken replied:
(To be clear, I am not saying you should do the followup - could be anyone.)

aheejin commented on src/wasm-interpreter.h:
```diff
@@ -1418,7 +1428,26 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+      trap("null ref");
+    }
+    const auto& exnData = exnref.getExnData();
+    WasmException exn;
+    exn.tag = exnData->tag;
+    for (auto item : exnData->payload) {
+      exn.values.push_back(item);
+    }
+    throwException(exn);
```
```suggestion
    throwException(exnData);
```
This can simplified if we change `WasmException` to take `ExnData` literal. See my comment on line 49-53.

aheejin reviewed (commented):
Thank you for the work! Haven't checked the tests yet, but the comments for the code part:

aheejin commented on test/lit/passes/coalesce-locals-eh-legacy.wast:
```diff
@@ -13,10 +15,6 @@
     (i32.const 1984)
   )
 
- (tag $e)
-
- (tag $any (param (ref any)))
-
```
I wonder what caused this change?

kripken replied:
Yeah, thanks, that's a good change I think.

aheejin commented on test/lit/passes/code-pushing-eh-legacy.wast:
```diff
@@ -7,69 +7,6 @@
   ;; CHECK:      (tag $e (param i32))
   (tag $e (param i32))
 
-  ;; CHECK:      (func $cannot-push-past-call (type $0)
```
Why does implementing the new EH delete the old legacy tests?

aheejin replied:
Sounds good!

aheejin reviewed (commented):
Haven't read the tests yet, but I wonder why the old `-legacy` tests have changed or deleted?

sjrd reviewed (commented):
Thanks @aheejin for the review. I addressed everything except the change to `WasmException`. I haven't managed to make it work yet. I left a comment with a link to my attempts.

sjrd commented:
> please see that at least a few thousand fuzzer iterations pass without issue here.

I let it run for more than 12,000 iterations without issue.

kripken commented on src/passes/Vacuum.cpp:
```diff
@@ -435,6 +435,22 @@ struct Vacuum : public WalkerPass<ExpressionStackWalker<Vacuum>> {
+      return;
+    }
+
+    // TODO Should we attempt the same thing as in `visitTry` when the body
+    // only throws, and there is a catch_all? We cannot `nop` the `TryTable`,
+    // though. Instead, we should turn it into an unconditional `br` to the
+    // target of the `catch_all`. Also, for a `catch_all_ref`, we cannot do that
+    // optimization at all since we would not know how to create the correct
+    // `exnref`. Note: This optimization may make more sense in RemoveUnusedBrs.
```
@aheejin The first check on `!throws` is on `curr->body`. The second check on `!hasUnremovableSideEffects` is on `curr`, which contains the body but also the try itself. And if the body can only throw then the entire try may have no effects.

aheejin replied:
@kripken Thanks, that makes sense.
@sjrd Yeah I still think we can remove it, given that it will only work for `catch_all` and I don't expect it to be used much, at least not in C++.

kripken commented on src/wasm-type.h:
```diff
@@ -388,6 +389,7 @@ class HeapType {
   bool isContinuation() const { return getKind() == HeapTypeKind::Cont; }
   bool isStruct() const { return getKind() == HeapTypeKind::Struct; }
   bool isArray() const { return getKind() == HeapTypeKind::Array; }
+  bool isExn() const { return isMaybeShared(HeapType::exn); }
```
@aheejin Do you mean shared as in `shared_ptr`? If not then I don't follow you, sorry.

The meaning here is the shared memory proposal. We do support shared-memory exn,

https://github.com/WebAssembly/binaryen/blob/9ce4b8a47b15c324b170f7572829532b7e5bfe47/test/lit/basic/shared-types.wast#L68

aheejin replied:
@sjrd What I meant was it cannot form cycles without GC enabled; sorry for not being clear. But I guess I misunderstood the meaning of the code anyway.

@kripken Thanks! I wasn't aware about that.

`shared_ptr` can be a separate issue. Actually we used to have this as a `unique_ptr`, before we removed `exnref` (https://github.com/WebAssembly/binaryen/blob/ca16f7cdc57549333474d7042128d2007bc35fad/src/literal.h#L44-L45). But I think in practice it wouldn't matter whether it is a `shared_ptr` or `unique_ptr` here.

aheejin commented on scripts/fuzz_opt.py:
```diff
@@ -1643,6 +1656,9 @@ def get_random_opts():
                 print('avoiding --flatten due to multivalue + reference types not supporting it (spilling of non-nullable tuples)')
                 print('TODO: Resolving https://github.com/WebAssembly/binaryen/issues/4824 may fix this')
                 continue
+            if '--enable-exception-handling' in FEATURE_OPTS:
+                print('avoiding --flatten due to exception-handling not supporting it (requires blocks with results)')
+                continue
```
This is going to disable `--flatten` even for the old EH, but unless we give the new EH a separate feature flag, which I'm not sure we should do, we can't help, I guess..?

kripken replied:
Good catch @aheejin , I missed that. But it seems reasonable to do, and not worth adding a new feature flag.

aheejin commented on src/wasm-interpreter.h:
```diff
@@ -4046,6 +4079,31 @@ class ModuleRunnerBase : public ExpressionRunner<SubType> {
+            }
+          }
+          if (curr->catchRefs[i]) {
+            ret.values.push_back(self()->makeExnData(e.tag, e.values));
```
Ping 😀

sjrd replied:
I did that in the branch that changes `WasmException` but in which I'm stuck:
https://github.com/sjrd/binaryen/compare/full-support-try-table...sjrd:binaryen:exnref-in-wasmexception
I don't think I can apply this particular suggestion without changing `WasmException`, can I?

aheejin commented on src/ir/linear-execution.h:
```diff
@@ -171,6 +171,11 @@ struct LinearExecutionWalker : public PostWalker<SubType, VisitorType> {
+        self->pushTask(SubType::doVisitTryTable, currp);
+        self->pushTask(SubType::doNoteNonLinear, currp);
+        break;
+      }
```
Does any test cover this?

sjrd replied:
Now yes: `simplify-locals-eh.wast` covers this.

aheejin commented:
> I also completed the translation of `-eh-legacy` lit tests to the new EH under `-eh` tests. That uncovered one incorrect optimization in `CodeFolding`, and two missed opportunities in `CodePushing` and `Vacuum`. If fixed those in the process.

Are the incorrect optimization and and missed opportunities about the old EH or the new EH? The new EH's optimizations have not really been implemented anyway prior to this PR so I guess they are about the old EH, but I can't find anything in the PR that fixes them. Can you elaborate?

sjrd commented on src/wasm/literal.cpp:
```diff
@@ -135,6 +141,9 @@ Literal::Literal(const Literal& other) : type(other.type) {
           gcData = other.gcData;
           return;
+        case HeapType::exn:
+          new (&exnData) std::shared_ptr<ExnData>(other.exnData);
```
So this was the culprit of the segmentation faults. Previously it read:
```c
exnData = other.exnData;
```
but that would execute the destructor of `shared_ptr<ExnData>` on the left-hand-side upon *construction* of this `Literal`, which caused segfaults.

I had copied the approach from `gcData = other.gcData;` 2 lines above. Is that one flawed as well? Should I change it to a `new` as well? Is that other one covered by any test?

kripken replied:
Hmm, this is at the limits of my C++ expertise, but I believe your new code is right, and the existing gcData code is wrong, and should be fixed. Let me look more into that separately (best to fix it in a separate PR, and I want to find out why tests did not fail - they should have).

kripken replied:
Ah, there is little mystery why that bad code did not crash... it was never reached, because we handle that case a few lines above :smile: I'll remove it.

https://github.com/WebAssembly/binaryen/blob/4c752b3d5ffecd6dc28fa1c0d15295bc0be60912/src/wasm/literal.cpp#L125-L127

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (return_call $foo-i32)
```
Can you add a line of comment on why this can be optimized, both here and the `eh-legacy` one, while you are at this? (`return_call` means calling the callee after we return the current function, so it will not be caught by `try`/`try_catch`, so we can optimize it)

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+      (block $tryend
+        (block $catch
+          (try_table (catch_all $catch)
+            (drop (i32.const 1))
```
Can you add a one-line comment saying expressions that cannot throw can be taken out of the try/try_table scope, both in here and the `eh-legacy` tests?

aheejin commented on test/lit/exec/eh.wast:
```diff
@@ -0,0 +1,45 @@
+  )
+
+  ;; CHECK:      [fuzz-exec] calling try-catch
+  (func $try-catch (export "try-catch")
```
```suggestion
  (func $try_table-catch (export "try_table-catch")
```

aheejin commented on test/lit/exec/eh.wast:
```diff
@@ -0,0 +1,45 @@
+
+  ;; CHECK:      [fuzz-exec] calling catchless-try
+  ;; CHECK-NEXT: [exception thrown: e-i32 3]
+  (func $catchless-try (export "catchless-try")
```
```suggestion
  (func $catchless-try_table (export "catchless-try_table")
```

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+  ;; CHECK-NEXT:   (i32.const 0)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-success (result i32)
```
```suggestion
  (func $try_table-call-optimize-terminating-tails-success (result i32)
```

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails (result i32)
```
```suggestion
  (func $try_table-call-optimize-terminating-tails (result i32)
```

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-call-return (result i32)
```
```suggestion
  (func $try_table-call-optimize-terminating-tails-call-return (result i32)
```

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (return_call $foo-i32)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-return-call (result i32)
```
```suggestion
  (func $try_table-call-optimize-terminating-tails-return-call (result i32)
```

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-expression-tails-success
```
```suggestion
  (func $try_table-call-optimize-expression-tails-success
```

aheejin commented on test/lit/passes/code-folding-eh.wast:
```diff
@@ -0,0 +1,291 @@
+  ;; CHECK-NEXT:   (unreachable)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-expression-tails
```
```suggestion
  (func $try_table-call-optimize-expression-tails
```

aheejin commented on test/lit/passes/code-pushing-eh.wast:
```diff
@@ -0,0 +1,300 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $can-push-past-try
```
```suggestion
  (func $can-push-past-try_table
```

aheejin commented on test/lit/passes/code-pushing-eh.wast:
```diff
@@ -0,0 +1,300 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $cannot-push-past-try
```
```suggestion
  (func $cannot-push-past-try_table
```

aheejin commented on test/lit/passes/dce-eh.wast:
```diff
@@ -0,0 +1,145 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $foo)
+  ;; CHECK-NEXT: )
+  (func $try_unreachable
```
```suggestion
  (func $try_table_unreachable
```

aheejin commented on test/lit/passes/dce-eh.wast:
```diff
@@ -0,0 +1,145 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+;; RUN: wasm-opt %s --dce -all -S -o - | filecheck %s
+
+;; If either try body or catch body is reachable, the whole try construct is
+;; reachable.
```
```suggestion
;; If either try_table body or any of catch handler is reachable, the whole
;; try_table construct is reachable.
```

aheejin commented on test/lit/passes/dce-eh.wast:
```diff
@@ -0,0 +1,145 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $throw
```
Given that this test doesn't contain any new instructions, I think we can delete this from `dce-eh-legacy.wast`, as we did for `code-pushing-eh-legacy.wast`.

aheejin commented on test/lit/passes/dce-eh.wast:
```diff
@@ -0,0 +1,145 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $throw-ref
```
```suggestion
  (func $throw_ref
```
`throw_ref` is an instruction name, so let's keep it that way

aheejin commented on test/lit/passes/code-pushing-eh.wast:
```diff
@@ -0,0 +1,300 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $cannot-push-past-throw-ref-within-catch
```
```suggestion
  (func $cannot-push-past-throw_ref-within-catch
```
`throw_ref` is an instruction name so let's keep it that way

aheejin commented on src/wasm-interpreter.h:
```diff
@@ -4046,6 +4068,32 @@ class ModuleRunnerBase : public ExpressionRunner<SubType> {
+          return ret;
+        }
+      }
+      // This exception is not caught by this try-catch. Rethrow it.
```
```suggestion
      // This exception is not caught by this try_table. Rethrow it.
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -235,15 +230,14 @@
+    (block $tryend
+      (try_table (catch_all $tryend)
         ;; This call cannot be optimized out, as the target throws. However, the
         ;; surrounding try-catch can be removed even without global effects
```
```suggestion
        ;; surrounding try_table can be removed even without global effects
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -235,15 +230,14 @@
         ;; This call cannot be optimized out, as the target throws. However, the
         ;; surrounding try-catch can be removed even without global effects
         ;; because the throw from the return_call is never observed by this
         ;; try-catch.
```
```suggestion
        ;; try_table.
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -258,8 +252,8 @@
+    (block $tryend
+      (try_table (catch_all $tryend)
         ;; This call cannot be optimized out, as the target may throw. However,
         ;; the surrounding try-catch can be removed even without global effects
```
```suggestion
        ;; the surrounding try_table can be removed even without global effects
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -283,8 +276,8 @@
+    (block $tryend
+      (try_table (catch_all $tryend)
         ;; This call cannot be optimized out, as the target may throw. However,
         ;; the surrounding try-catch can be removed even without global effects
```
```suggestion
        ;; the surrounding try_table can be removed even without global effects
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -182,49 +182,44 @@
         ;; This call cannot be optimized out, as the target throws. However, the
-        ;; entire try-catch can be, since the call's only effect is to throw,
-        ;; and the catch_all catches that.
+        ;; entire try-catch could be, since the call's only effect is to throw,
```
```suggestion
        ;; entire try_table could be, since the call's only effect is to throw,
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -293,84 +286,71 @@
-      (do
+    (block $tryend
+      (try_table (catch_all $tryend)
         ;; Even though the body of the previous function is a try-catch_all, the
```
```suggestion
        ;; Even though the body of the previous function has a catch_all, the
```
By the way I'm not sure what the "previous function" here (and the original test) refers to. @kripken

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -293,84 +286,71 @@
         ;; Even though the body of the previous function is a try-catch_all, the
         ;; function still throws because of its return_call, so this cannot be
-        ;; optimized out, but once again the entire try-catch can be.
+        ;; optimized out, but once again the entire try-catch could be. Again,
```
```suggestion
        ;; optimized out, but once again the entire try_table could be. Again,
```

aheejin commented on test/lit/passes/global-effects.wast:
```diff
@@ -435,16 +408,13 @@
   ;; INCLUDE-NEXT:  )
   ;; INCLUDE-NEXT: )
   (func $call-throw-or-unreachable-and-catch (param $x i32)
     ;; This try-catch-all's body will either call a throw or an unreachable.
```
```suggestion
    ;; This try_table's body will either call a throw or an unreachable.
```

aheejin commented on test/lit/passes/gufa-eh.wast:
```diff
@@ -0,0 +1,57 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $try-table-target-block-is-not-unreachable (result i32)
```
```suggestion
  (func $try_table-target-block-is-not-unreachable (result i32)
```
`try_table` is an instruction name

aheejin commented on test/lit/passes/gufa-eh.wast:
```diff
@@ -0,0 +1,57 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-table-materializes-exnref
```
```suggestion
  (func $try_table-materializes-exnref
```

aheejin commented on test/lit/passes/local-subtyping.wast:
```diff
@@ -568,4 +571,50 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-table-catch-result
```
```suggestion
  (func $try_table-catch-result
```

aheejin commented on test/lit/passes/local-subtyping.wast:
```diff
@@ -568,4 +571,50 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try-table-ref
```
```suggestion
  (func $try_table-ref
```

aheejin commented on test/lit/passes/simplify-locals-eh.wast:
```diff
@@ -0,0 +1,225 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $call-cannot-be-sinked-into-try (local $0 i32)
```
```suggestion
  (func $call-cannot-be-sinked-into-try_table (local $0 i32)
```

aheejin commented on test/lit/passes/simplify-locals-eh.wast:
```diff
@@ -0,0 +1,225 @@
+  ;; CHECK-NEXT: )
+  (func $call-cannot-be-sinked-into-try (local $0 i32)
+    (drop
+      ;; This local.tee should NOT be sinked into 'try' below, because it may
+      ;; throw
```
```suggestion
      ;; This local.tee should NOT be sinked into 'try_table' below, because it
      ;; may throw
```

aheejin commented on test/lit/passes/simplify-locals-eh.wast:
```diff
@@ -0,0 +1,225 @@
+  ;; CHECK-NEXT: )
+  (func $non-call-can-be-sinked-into-try (local $0 i32)
+    (drop
+      ;; This local.tee can be sinked into 'try' below, because it cannot throw
```
```suggestion
      ;; This local.tee can be sinked into 'try_table' below, because it cannot
      ;; throw
```

aheejin commented on test/lit/passes/simplify-locals-eh.wast:
```diff
@@ -0,0 +1,225 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $non-call-can-be-sinked-into-try (local $0 i32)
```
```suggestion
  (func $non-call-can-be-sinked-into-try_table (local $0 i32)
```

aheejin commented on test/lit/passes/simplify-locals-eh.wast:
```diff
@@ -0,0 +1,225 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $return-call-can-be-sinked-into-try (result i32)
```
```suggestion
  (func $return-call-can-be-sinked-into-try_table (result i32)
```

aheejin commented on test/lit/passes/simplify-locals-eh.wast:
```diff
@@ -0,0 +1,225 @@
+        (if (result i32)
+          (i32.const 0)
+          (then
+            (return_call $return-call-can-be-sinked-into-try)
```
```suggestion
            (return_call $return-call-can-be-sinked-into-try_table)
```

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 2)
+  ;; CHECK-NEXT: )
+  (func $inner-try-catch_all-test (result i32)
```
```suggestion
  (func $inner-try_table-catch_all-test (result i32)
```

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+  ;; CHECK-NEXT: )
+  (func $inner-try-catch_all-test (result i32)
+    (local $0 i32)
+    ;; The exception thrown in the inner try is caught by the inner catch_all,
+    ;; so the outer try body does not throw and the outer try-catch can be
+    ;; removed
```
```suggestion
    ;; The exception thrown in the inner try_table is caught by the inner
    ;; catch_all, so the outer try_table body does not throw and the outer
    ;; try_table can be removed
```

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $trivial-catch-all-of-throw (local $0 i32)
+    ;; This try-catch's body throws, but the catch-all catches it, so the entire
+    ;; try could be optimized out. We do this for `try` but not yet for
+    ;; `try_table`.
```
```suggestion
    ;; This try_table's body throws, but the catch_all catches it, so the entire
    ;; try_table could be optimized out. We do this for `try` but not yet for
    ;; `try_table`.
```

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+    (block $catch
+      (try_table (catch_all $catch)
+        ;; This returns before it throws, so we can optimize out the surrounding
+        ;; try-catch.
```
```suggestion
        ;; try_table.
```

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+    (block $catch
+      (try_table (catch_all $catch)
+        ;; This returns before it throws, so we can optimize out the surrounding
+        ;; try-catch.
```
```suggestion
        ;; try_table.
```

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,253 @@
+    (block $catch
+      (try_table (catch_all $catch)
+        ;; This returns before it throws, so we can optimize out the surrounding
+        ;; try-catch.
```
```suggestion
        ;; try_table.
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    (throw $e-i32-f32 (i32.const 3) (f32.const 3.5))
+  )
+
+  (func (export "try_nothrow") (result i32)
```
```suggestion
  (func (export "try_table_nothrow") (result i32)
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_throw_catch") (result i32)
```
```suggestion
  (func (export "try_table_throw_catch") (result i32)
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_throw_nocatch") (result i32)
```
```suggestion
  (func (export "try_table_throw_nocatch") (result i32)
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_throw_catchall") (result i32)
```
```suggestion
  (func (export "try_table_throw_catchall") (result i32)
```

aheejin commented on test/spec/return_call_eh.wast:
```diff
@@ -1,34 +1,31 @@
+    (block $catch
+      (return
+        (try_table (result i32) (catch_all $catch)
+          (call $return-call-in-try)
```
```suggestion
          (call $return-call-in-try_table)
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_call_catch") (result i32)
```
```suggestion
  (func (export "try_table_call_catch") (result i32)
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_throw_multivalue_catch") (result i32) (local $x (tuple i32 f32))
```
```suggestion
  (func (export "try_table_throw_multivalue_catch") (result i32) (local $x (tuple i32 f32))
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_throw_rethrow") (local $x (tuple i32 exnref))
```
```suggestion
  (func (export "try_table_throw_throw_ref") (local $x (tuple i32 exnref))
```

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+    )
+  )
+
+  (func (export "try_call_rethrow")
```
```suggestion
  (func (export "try_table_call_throw_ref")
```

aheejin reviewed (commented):
Mostly test names and comments nits. I think you can "Add suggestion to batch" on the ones you'd like to accept and batch-commit them, without manually fixing them.

sjrd commented on src/ir/possible-contents.h:
```diff
@@ -473,6 +473,12 @@ struct TagLocation {
+// of a try_table.
+struct CaughtExnRefLocation {
+  bool operator==(const CaughtExnRefLocation& other) const { return true; }
+};
```
1. Yes, there is basically a unique source of non-null `exnref`s. This is similar to `getNullLocation`, which creates a unique source of zero values per type.
2. I don't think we can, because at
  https://github.com/WebAssembly/binaryen/blob/cde2a52fc6d679d53d1c5b72b081992d53e01c6b/src/ir/possible-contents.cpp#L2116-L2117
  the values associated to roots are put in an `st::map<Location, PossibleContents>`, without trying to `combine` them. That means each Location can only accommodate a single root value. If we use a `TagLocation` or even the target `BreakLocation` and assign them a root value, we can enter into conflict with other things that would try to put a different root value (such as a null value) for the same location.

kripken replied:
I think I see now, thanks. So we just need **a** source of content for the places that read a thrown exnref.

aheejin replied:
It's been a while since I read GUFA code so my understanding is bad, but why are all `exnref`s treated the same? All `null`s are actually the same value and can be used interchangeably, but each `exnref` contains a different payload and tag, no?

sjrd commented:
Overnight I ran the fuzzer while *allowing* it to use the new EH sources as inputs. I did not discover any issue within wasm-opt, but I think I found a bug in the V8 validator (after ~6,000 iterations). It refuses to validate the following module:
```wat
(module
 (type $0 (func))
 (func $0
  (throw_ref
   (ref.null noexn)
  )
 )
)
```
with the error
```
/localhome/doeraene/projects/binaryen/scripts/fuzz_shell.js:155: CompileError: WebAssembly.Module(): Compiling function #0 failed: invalid type for throw_ref: expected exnref, found noexnref @+25
var module = new WebAssembly.Module(binary);
             ^
CompileError: WebAssembly.Module(): Compiling function #0 failed: invalid type for throw_ref: expected exnref, found noexnref @+25
    at /localhome/doeraene/projects/binaryen/scripts/fuzz_shell.js:155:14
```
I don't think this is correct, because `noexnref` is a subtype of `exnref`. It should allow to validate the `throw_ref` (even if it will inevitably `trap` at run-time).

It looks similar to https://issues.chromium.org/issues/332931390 which I had already reported some time ago.

kripken commented:
> `invalid type for throw_ref: expected exnref, found noexnref @+25`

That does look like a bug. Until v8 fixes it, we'll need to disable fuzzing of exnref. That means disabling all exceptions testing in v8 since we don't have separate feature flags, which is somewhat unfortunate, but we will still be fuzzing it in our interpreter so it's not too bad.

To disable it in V8 only, see

https://github.com/WebAssembly/binaryen/blob/d5675780f45f4e2994787fd7dbc5283a6c0162f0/scripts/fuzz_opt.py#L840-L844

kripken commented on src/ir/possible-contents.h:
```diff
@@ -473,6 +473,12 @@ struct TagLocation {
 };
 
+// The location of an exnref materialized by a catch_ref or catch_all_ref clause
+// of a try_table.
```
```suggestion
// of a try_table. No data is stored here (exnrefs contain a stack at runtime, but we don't track that), so this is the same as NullLocation in a way: we just need *a* source of contents for places that receive an exnref.
```

aheejin replied:
It's been a while since I read GUFA code so my understanding is bad, but why are all exnrefs treated the same? All nulls are actually the same value and can be used interchangeably, but each exnref contains a different payload and tag, no?

kripken replied:
Isn't the payload separate from the exnref? When we catch with an exnref we jump to a block that gets the payload as multivalues + an exnref at the end IIUC.

sjrd replied:
GUFA doesn't do any abstract interpretation, AFAICT. It only tracks specific literals, and unknown things of certain types. So all unknown `(ref exn)` are the same anyway.

aheejin replied:
`exnref` conceptually contains the payload, even though it is opaque, no? I think we can treat it in the same way we treat `externref` or `anyref`, given that they also contain something opaque. How do we treat them in GUFA? Do we treat all of them as the same literal?

aheejin replied:
In the new EH (exnref) scheme, when we rethrow an exnref using `throw_ref`, we rethrow the whole package, including a tag and a payload. So `exnref` contains both.

kripken replied:
> In the new EH (exnref) scheme, when we rethrow an exnref using throw_ref, we rethrow the whole package, including a tag and a payload. So exnref contains both. 

Oh, I didn't realize that. So when we catch, we are sending the payload as values, and also sending it wrapped in the `exnref` object (which also has the stack and tag id)? So we send it both wrapped and unwrapped?

>  I think we can treat it in the same way we treat externref or anyref, given that they also contain something opaque.

Makes sense, yeah. That is basically what the code does. E.g. if an import returns an anyref, then we know nothing about it but the type, so we fill that `ResultLocation` (result of a function) with `addRoot`, which just sets it to an unknown value of that type. The code here does the same, but we do need a Location to read the exnref from (it isn't a `ResultLocation` or anything else we have, yet), so we add that as well.

aheejin replied:
> > In the new EH (exnref) scheme, when we rethrow an exnref using throw_ref, we rethrow the whole package, including a tag and a payload. So exnref contains both.
> 
> Oh, I didn't realize that. So when we catch, we are sending the payload as values, and also sending it wrapped in the `exnref` object (which also has the stack and tag id)? So we send it both wrapped and unwrapped?

No I think, at least conceptually (because the actual implementation in the engines can differ), we just send a package (=exnref). And when caught, depending on the kind of `catch` clauses, we provide the `exnref` as is or extract values from it and provide the values.

> > I think we can treat it in the same way we treat externref or anyref, given that they also contain something opaque.
> 
> Makes sense, yeah. That is basically what the code does. E.g. if an import returns an anyref, then we know nothing about it but the type, so we fill that `ResultLocation` (result of a function) with `addRoot`, which just sets it to an unknown value of that type. The code here does the same, but we do need a Location to read the exnref from (it isn't a `ResultLocation` or anything else we have, yet), so we add that as well.

Still not sure why we need a separate location kind, given that `externref`/`anyref` doesn't have any. `exnref` here is just a return value from a `block`, so it can be treated as such, no? Do we need a separate `Location` kind for every kind of block return value?

kripken replied:
The issue is, where does the block return value receive the value from? `anyref` arrives from e.g. a function result, which we already have a Location type for. But here we are not just adding a new type but also a new control flow construct that sends that value.

But maybe we could instead add a root on the target of the branch, though, is that what you mean? I suppose there is just one such target, and we know exactly what it is. I think that would be safe and correct to do here (that location can't be marked as a root for any other reason), but it does seem a little cleaner to me to keep the code as it is, since the branch target is not conceptually a "root" (e.g. it may receive values from other branches than this, so it needs to combine them all. We do know that this unknown value will "win", but conceptually we are still combining there).

aheejin replied:
Hmm I see. Then I guess even though all `exnref`s are treated as the same constant it wouldn't matter in GUFA because it doesn't do anything with constants, right?

kripken replied:
Yeah, we just move constants around, but don't do anything with the values otherwise. So this should be ok as is, I think.

sjrd replied:
> Hmm I see. Then I guess even though all `exnref`s are treated as the same constant it wouldn't matter in GUFA because it doesn't do anything with constants, right?

They're not treated like the same *constant*. They're all treated as some *unknown* value of type `(ref exn)`. GUFA makes a difference between actual constants and unknown values of some type. If we did register them as the same constant, that would be bad, as GUFA would then propagate that constant at all destinations.

kripken commented on src/ir/possible-contents.h:
```diff
@@ -608,6 +615,10 @@ template<> struct hash<wasm::TagLocation> {
 };
 
+template<> struct hash<wasm::CaughtExnRefLocation> {
+  size_t operator()(const wasm::CaughtExnRefLocation& loc) const { return 0; }
```
0 is actually risky, as if we had other such things then they'd also use 0 in the same natural way, and all collide. How about `std::hash<const char*>()("caught-exnref-location")` - ? That hashes the pointer, and that pointer should be unique as other things wouldn't use that text presumably.

kripken commented on src/ir/possible-contents.cpp:
```diff
@@ -1134,8 +1134,38 @@ struct InfoCollector
+      }
+
+      if (curr->catchRefs[tagIndex]) {
+        Location location = CaughtExnRefLocation{};
```
```suggestion
        auto location = CaughtExnRefLocation{};
```

aheejin commented on test/lit/passes/local-subtyping.wast:
```diff
@@ -568,4 +571,50 @@
+  ;; CHECK-NEXT: )
+  (func $try_table-catch-result
+    (drop
+      ;; Must not be refined to nullref.
+      ;; An anyref comes from the tag's payload because it is not a catch_all.
+      (block $catch (result anyref)
```
Does this mean that this `(result anyref)` should not be refined to something more refined than `anyref`? Does `LocalSubtyping` refine `block`'s return types too? (I'm not familiar with the pass)

And I think `it is not a catch_all` in the comment may not be necessary.

sjrd replied:
> Does this mean that this `(result anyref)` should not be refined to something more refined than `anyref`?

Yes, that's it. I specifically added "(result nullref)" in the comment.

> Does `LocalSubtyping` refine `block`'s return types too? (I'm not familiar with the pass)

It does through `ReFinalize`. It's a bit weird to me that `ReFinalize::visitNode` and `Node::finalize` are separate but seem to do very similar things. For some passes, `finalize` is what matters; for others (like `LocalSubtyping`), it's `ReFinalize`.

aheejin commented on test/lit/passes/vacuum-eh.wast:
```diff
@@ -0,0 +1,254 @@
+    )
+  )
+
+  ;; CHECK:      (func $br-in-catch (type $void)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (unreachable)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $br-in-catch
+    ;; When the try_table is removed, the removal of the implicit branch to the
+    ;; catch target should be registered so that its type will be correctly
+    ;; updated to unreachable.
+    (drop
+      (block $catch (result i32)
+        (try_table (catch $e $catch)
+          (unreachable)
+        )
+        (unreachable)
+      )
+    )
+  )
+
```
```suggestion
```

The original test has a `br` within the `catch` body (hence the name) but this doesn't. And that test was added in #2765, which fixed a bug that `typeUpdater.noteRecursiveRemoval(curr->catchBody);` was missing for the catch body. Given that we don't have 'catch bodies' anymore and don't need that `noteRecursiveRemoval`, I think this test can be removed.

aheejin commented on test/spec/exception-handling.wast:
```diff
@@ -1,3 +1,153 @@
+            )
+          )
+        )
+        (br $tryend (i32.const 3))
```
```suggestion
        (br $tryend (i32.const 4))
```
It'd be better to make each constant different (the original test has 4 here too)

sjrd commented:
I won't have a computer until Tuesday, do I'll come back to this then. 

For the conflicts, should I resolve them by merging main into this branch, or rebasing on top of main?

kripken commented:
@sjrd Merging would be better, thanks.

sjrd commented on test/lit/passes/gufa-eh.wast:
```diff
@@ -0,0 +1,57 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $try_table-materializes-exnref
```
I elaborated in a comment:

> If we do not connect a non-null exnref value to the branch target, GUFA will think no value can possibly get out of that block, and will insert an unreachable instruction after the block.

sjrd commented:
With the latest changes, and with disabling the following lines from the fuzzer (see [comments here](https://github.com/WebAssembly/binaryen/commit/b3e22d29451fbf52521d59ea34e8a5d735c4149c#diff-9169a8b6e3b7062e784ef435b54381c119ee3c644d44ee83a05efdf0d64a472e)):
https://github.com/WebAssembly/binaryen/blob/2c9c74d8b64e1776c6c374af8631995b0be606f1/src/tools/fuzzing/fuzzing.cpp#L3304-L3310
I could successfully run 20,000+ iterations of the fuzzer on this PR, when also allowing the fuzzer to take the new EH tests as inputs.

@kripken Do you think that's good enough to remove the following lines:
https://github.com/WebAssembly/binaryen/pull/6814/files#diff-3c641dc74f54559f1df2b85afcac1c8e199a33edfbbf80a481167ad2e1b01f90L355-R372
?

kripken commented:
> I could successfully run 20,000+ iterations of the fuzzer on this PR, when also allowing the fuzzer to take the new EH tests as inputs. Do you think that's good enough to remove the following lines

Nice! Yes, that seems enough. Likely the fuzzer will eventually find more issues but so long as there is nothing known and high-frequency that fails, that's what we want.

kripken reviewed (approved):
lgtm % any remaining questions @aheejin has, and to enable the fuzzer as just discussed.

sjrd commented:
Thanks. Fuzzer enabled.

aheejin reviewed (approved):
Thank you for the work!

kripken commented:
Thanks from me as well @sjrd , this is great!
